### PR TITLE
config_tool: update cpu affinity data load and check

### PR DESCRIPTION
**Pull request by Weiyi-Feng**
config_tool: add cpu affinity check
1. ensure cpu affinity in launch xml is subset of its scenario settings.
2. cpu_affinity is a must have parameter for acrn_dm, if the user did not provide cpu affinity in launch xml, load it from scenario settings.

[External_System_ID] ACRN-7453
Tracked-On: #6724
Signed-off-by: Weiyi Feng <weiyix.feng@intel.com>

### CODE REVIEW

**Comment by wenlingz** [comment]
@Weiyi-Feng build fail, please refer:
1636350745/build//hypervisor/include/config.h generated
| Board configurations for ehl-crb-b is generated successfully.
| Traceback (most recent call last):
|   File "/home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6785-1636350745/git/hypervisor/../misc/config_tools/scenario_config/scenario_cfg_gen.py", line 312, in <module>
|     err_dic = main(ARGS)
|   File "/home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6785-1636350745/git/hypervisor/../misc/config_tools/scenario_config/scenario_cfg_gen.py", line 250, in main
|     (err_dic, scenario_items) = validate_scenario_setting(params['--board'], params['--scenario'])
|   File "/home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6785-1636350745/git/hypervisor/../misc/config_tools/scenario_config/scenario_cfg_gen.py", line 181, in validate_scenario_setting
|     vm_info.check_item()
|   File "/home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6785-1636350745/git/misc/config_tools/scenario_config/scenario_item.py", line 402, in check_item
|     err_dic = scenario_cfg_lib.vm_cpu_affinity_check(self.scenario_info, self.cpus_per_vm, "pcpu_id")
|   File "/home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6785-1636350745/git/misc/config_tools/scenario_config/../library/scenario_cfg_lib.py", line 352, in vm_cpu_affinity_check
|     for vm_id, cpu_ids in cpu_affinity.items():
| AttributeError: 'str' object has no attribute 'items'
| Start pre-build static check ...
| make -C ../misc/hv_prebuild BOARD=ehl-crb-b SCENARIO=hybrid_rt TARGET_DIR=/home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6785-1636350745/build//hypervisor/configs
| make[1]: Entering directory '/home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6785-1636350745/git/misc/hv_prebuild'
| make[1]: *** No rule to make target '/home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6785-1636350745/build//hypervisor/configs/scenarios/hybrid_rt/vm_configurations.c', needed by 'default'.  Stop.
| make[1]: Leaving directory '/home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6785-1636350745/git/misc/hv_prebuild'
| Makefile:412: recipe for target 'pre_build' failed
| make: *** [pre_build] Error 2
| make: Leaving directory '/home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6785-1636350745/git/hypervisor'
| ERROR: oe_runmake failed
| WARNING: /home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6785-1636350745/temp/run.do_compile.788:181 exit 1 from 'exit 1'
| WARNING: Backtrace (BB generated script):
| 	#1: bbfatal_log, /home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6785-1636350745/temp/run.do_compile.788, line 181
| 	#2: die, /home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6785-1636350745/temp/run.do_compile.788, line 162
| 	#3: oe_runmake, /home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6785-1636350745/temp/run.do_compile.788, line 151
| 	#4: do_compile, /home/meta-acrn/workspace/build/master-acrn-sos/work/x86_64-linux/acrn-hypervisor-native/6785-1636350745/temp/run.do_compile.788, line 146
| 	#5: main, /home/meta-acrn/workspace/build/master-acrn-sos/work/x8

File: misc/config_tools/library/scenario_cfg_lib.py [hunk 1]
```diff
@@ -347,37 +347,46 @@ def vm_cpu_affinity_check(config_file, id_cpus_per_vm_dic, item):
...
-                return err_dic
-            else:
-                use_cpus.append(cpu)
+    # validate cpu_affinity config with scenario file
+    scenario_cpu_aff = common.get_leaf_tag_map(scenario_file, "cpu_affinity", "pcpu_id")
+    for vm_id, cpu_ids in cpu_affinity.items():
+        for vm_cpu in cpu_ids:
+            if vm_cpu not in scenario_cpu_aff[vm_id]:
+                key = "vm:id={},{}".format(vm_id, 'pcpu_id')
+                err_dic[key] = "This pcpu is not configured in this vm's scenario file. Please re-configurate them!"
```
Comment (gvancuts) [inline misc/config_tools/library/scenario_cfg_lib.py:356]:
```suggestion
                err_dic[key] = "This pCPU is not included in this VM's allowed CPU pool. Please update your scenario file accordingly or remove it from this list."
```

File: misc/config_tools/library/scenario_cfg_lib.py [hunk 2]
```diff
@@ -347,37 +347,46 @@ def vm_cpu_affinity_check(config_file, id_cpus_per_vm_dic, item):
...
+
+    if err_dic:
+        return err_dic
+
+    for vm_i,cpu in cpu_affinity.items():
+        if cpu is not None and cpu in use_cpus and not cpu_sharing_enabled:
+            key = "vm:id={},{}".format(vm_i, 'pcpu_id')
+            err_dic[key] = "The same pcpu was configurated in <pcpu_id>/<cpu_affinity>, but CPU sharing is disabled by 'SCHED_NOOP'. Please re-configurate them!"
```
Comment (gvancuts) [inline misc/config_tools/library/scenario_cfg_lib.py:364]:
```suggestion
            err_dic[key] = "The same pCPU was configured in <pcpu_id>/<cpu_affinity>, but CPU sharing is disabled by 'SCHED_NOOP'. Please enable CPU sharing or update your CPU affinity configuration."
```

**Review by gvancuts — APPROVED** [review]

**Comment by wenlingz** [comment]
OK to verify

**Comment by wenlingz** [comment]
try verify EHL

**Review by NanlinXie — APPROVED** [review]

**Pull request merged**